File: typescript/data-structures/queue/queue.ts
namespace Queue {
  type Queue<T> = {
    max_capacity: number;
    values: T[];
  };

  const newQueue = <T>(max_capacity: number): Queue<T> => ({
    max_capacity,
    values: [],
  });

  const isEmpty = <T>(queue: Queue<T>): boolean => {
    return length(queue) === 0;
  };

  const length = <T>(queue: Queue<T>): number => {
    return queue.values.length;
  };

  const enqueue = <T>(value: T, queue: Queue<T>): Queue<T> => {
    if (queue.max_capacity == length(queue)) {
      return queue;
    }

    return { ...queue, values: [...queue.values, value] };
  };

  const dequeue = <T>(queue: Queue<T>): [T, Queue<T>] | [] => {
    if (isEmpty(queue)) {
      return [];
    }

    const [value, ...values] = queue.values;
    return [value, { ...queue, values }];
  };

  const peek = <T>(queue: Queue<T>): T | [] => {
    if (isEmpty(queue)) {
      return [];
    }

    const [value] = queue.values;
    return value;
  };
}

File: clang/data-structures/queues/queue.c
#include <limits.h>
#include <stdio.h>
#include <stdlib.h>

typedef struct Queue {
  int head, tail;
  int capacity;
  int *array;
  int size;
} Queue;

Queue *make_queue(int capacity) {
  struct Queue *queue = malloc(sizeof(struct Queue));

  queue->head = queue->tail = -1;
  queue->capacity = capacity;
  queue->array = malloc(queue->capacity * sizeof(int));
  queue->size = 0;

  return queue;
}

unsigned is_empty(Queue *queue) { return (queue->size == 0); }

unsigned is_full(Queue *queue) { return (queue->size == queue->capacity); }

void enqueue(Queue *queue, int data) {
  if (is_full(queue))
    return;

  queue->tail = (queue->tail + 1) % queue->capacity;
  queue->array[queue->tail] = data;

  if (queue->head == -1)
    queue->head = queue->tail;

  queue->size += 1;
}

int dequeue(Queue *queue) {
  if (is_empty(queue))
    return INT_MIN;

  int data = queue->array[queue->head];

  if (queue->head == queue->tail) {
    queue->head = queue->tail = -1;
    queue->size = 0;
  } else {
    queue->head = (queue->head + 1) % queue->capacity;
    queue->size -= 1;
  }

  return data;
}

void print_queue(Queue *queue) {
  int *arr = queue->array;
  int size = queue->size;

  for (int k = 0; k < size; k++)
    printf("%d ", arr[k]);
}

int main(void) {
  Queue *queue = make_queue(16);

  enqueue(queue, 0);
  enqueue(queue, 1);
  enqueue(queue, 2);
  enqueue(queue, 3);
  enqueue(queue, 4);
  enqueue(queue, 5);
  enqueue(queue, 6);
  enqueue(queue, 7);

  printf("dequeued: %d\n", dequeue(queue));
  printf("dequeued: %d\n", dequeue(queue));

  print_queue(queue);

  return 0;
}
